scarlet/fft.py

```diff
@@ -0,0 +1,297 @@
+        image = _centered(image, image_shape)
+        return Fourier(image, {tuple(fft_shape): image_fft}, axes)
+
+    @property
+    def image(self):
+        """The real space image"""
+        return self._image
+
+    @property
+    def axes(self):
+        """The axes that are transormed"""
+        return self._axes
+
+    @property
+    def shape(self):
+        """The shape of the real space image"""
+        return self.image.shape
+
+    def fft(self, fft_shape):
+        """The FFT of an image for a given `fft_shape`
+        """
+        fft_shape = tuple(fft_shape)
+        # If this is the first time calling `fft` for this shape,
+        # generate the FFT.
+        if fft_shape not in self._fft:
+            image = _pad(self.image, fft_shape, self._axes)
+            self._fft[fft_shape] = np.fft.rfftn(np.fft.ifftshift(image, self._axes), axes=self._axes)
+        return self._fft[fft_shape]
+
+    def __len__(self):
+        return len(self.image)
+
+    def normalize(self):
+        """Normalize the image to sum to one
+        """
+        if self._axes is not None:
+            indices = [slice(None)] * len(self.shape)
+            for ax in self._axes:
+                indices[ax] = None
+        else:
+            indices = [None] * len(self.shape)
+        indices = tuple(indices)
+        normalization = 1/self._image.sum(axis=self._axes)
+        self._image *= normalization[indices]
+        for shape, image_fft in self._fft.items():
+            self._fft[shape] *= normalization[indices]
+
+    def update_dtype(self, dtype):
+        if self.image.dtype != dtype:
+            self._image = self._image.astype(dtype)
+            for shape in self._fft:
+                self._fft[shape] = self._fft[shape].astype(dtype)
+
+    def sum(self, axis=None):
+        return self.image.sum(axis)
+
+    def max(self, axis=None):
+        return self.image.max(axis=axis)
+
+    def __getitem__(self, index):
+        # Make the index a tuple
+        if not hasattr(index, "__getitem__"):
+            index = tuple([index])
+
+        # Axes that are removed from the shape of the new object
+        removed = np.array([n for n, idx in enumerate(index) if not isinstance(idx, slice)])
+        # Only propagate the shape indices for axes not removed
+
+        # Only propagate axes that are sliced or not indexed and
+        # decrement them by the number of removed axes smaller than each one
+        axes = tuple([ax-np.sum(removed < ax) for ax in self.axes if ax not in removed])
+        # Create views into the fft transformed values, appropriately adjusting
+        # the shapes for the new axes
+        fft_kernels = {
+            tuple([s for idx, s in enumerate(shape) if self.axes[idx] not in removed]): kernel[index]
+            for shape, kernel in self._fft.items()
+        }
+        return Fourier(self.image[index], fft_kernels, axes=axes)
+
+
+def _kspace_operation(image1, image2, padding, op, shape):
```

Comment:
Yes that's true and if you prefer I can change the name. I still think of the `Fourier` instance as an image, since the class is really a wrapper of an image that includes convenience functions to operate on its Fourier transform. But I can see how this might be confusing.

Reply:
I don't mind the name all that much, but it should be clearer in the docstring, I think.

Reply:
Sure, no problem. I usually try to keep the docstrings light for private functions that are not part of the public API but in this case I think it is important to make it clear what is expected.